ICE: not yet implemented: FIXME(async_closures): Lower these to SMIR

<!--
[31mICE[0m: Rustc ./a.rs '-Zunpretty=stable-mir --edition=2018 -ooutputfile -Zdump-mir-dir=dir' 'thread 'rustc' panicked at compiler/rustc_smir/src/rustc_smir/convert/mir.rs:569:17: 'not yet implemented: FIXME(async_closures): Lower these to SMIR'', 'thread 'rustc' panicked at compiler/rustc_smir/src/rustc_smir/convert/mir.rs:569:17: 'not yet implemented: FIXME(async_closures): Lower these to SMIR''
File: /tmp/im/a.rs
-->

snippet:
````rust
//@compile-flags: -Zunpretty=stable-mir --edition=2018
use std::future::ready;
pub fn main() {
    let mut vec: Vec<String> = vec![];

    let closure = async || {
        vec.push(ready(String::from("")).await);
    };
}

````

Version information
````
rustc 1.85.0-nightly (f4f0fafd0 2024-12-13)
binary: rustc
commit-hash: f4f0fafd0c7849e162eddbc69fa5fe82dbec28c7
commit-date: 2024-12-13
host: x86_64-unknown-linux-gnu
release: 1.85.0-nightly
LLVM version: 19.1.5

````

Possibly related line of code:
https://github.com/rust-lang/rust/blob/f4f0fafd0c7849e162eddbc69fa5fe82dbec28c7/compiler/rustc_smir/src/rustc_smir/convert/mir.rs#L563-L575

Command:
`/home/matthias/.rustup/toolchains/master/bin/rustc -Zunpretty=stable-mir --edition=2018`

<details><summary><strong>Program output</strong></summary>
<p>

```
warning: unused variable: `closure`
 --> /tmp/icemaker_global_tempdir.AGr5o5x8v81C/rustc_testrunner_tmpdir_reporting.LjFuOhg1b1UO/mvce.rs:5:9
  |
5 |     let closure = async || {
  |         ^^^^^^^ help: if this is intentional, prefix it with an underscore: `_closure`
  |
  = note: `#[warn(unused_variables)]` on by default

thread 'rustc' panicked at compiler/rustc_smir/src/rustc_smir/convert/mir.rs:569:17:
not yet implemented: FIXME(async_closures): Lower these to SMIR
stack backtrace:
   0:     0x728a3016887a - <std::sys::backtrace::BacktraceLock::print::DisplayBacktrace as core::fmt::Display>::fmt::ha7f0c0a9782d7e72
   1:     0x728a30813ba6 - core::fmt::write::h349a90590cd44cc2
   2:     0x728a317c5e51 - std::io::Write::write_fmt::hb75740cea2350d81
   3:     0x728a301686d2 - std::sys::backtrace::BacktraceLock::print::hbaa671e034187daa
   4:     0x728a3016abea - std::panicking::default_hook::{{closure}}::he5b32cfe851d3684
   5:     0x728a3016aa33 - std::panicking::default_hook::hfb25fc6fe8d3f7df
   6:     0x728a2f2cb5c8 - std[4908e3f2635a52b0]::panicking::update_hook::<alloc[d0f40dee7a07bc41]::boxed::Box<rustc_driver_impl[da799d46c754177c]::install_ice_hook::{closure#0}>>::{closure#0}
   7:     0x728a3016b3a8 - std::panicking::rust_panic_with_hook::had4ab0340b9595a9
   8:     0x728a3016b066 - std::panicking::begin_panic_handler::{{closure}}::hfe00e9533ed00ed0
   9:     0x728a30168d19 - std::sys::backtrace::__rust_end_short_backtrace::h15687a7da048e988
  10:     0x728a3016ad5d - rust_begin_unwind
  11:     0x728a2cdb9920 - core::panicking::panic_fmt::h67b4e51ad0c096d3
  12:     0x728a2fe2d8a6 - <rustc_middle[925dc401787dd4f]::mir::syntax::AggregateKind as rustc_smir[2e64afc0fcf6e61f]::rustc_smir::Stable>::stable
  13:     0x728a2fe2c737 - <rustc_middle[925dc401787dd4f]::mir::syntax::Rvalue as rustc_smir[2e64afc0fcf6e61f]::rustc_smir::Stable>::stable
  14:     0x728a2fe2c51e - <rustc_middle[925dc401787dd4f]::mir::syntax::StatementKind as rustc_smir[2e64afc0fcf6e61f]::rustc_smir::Stable>::stable
  15:     0x728a2fe2ba58 - <rustc_middle[925dc401787dd4f]::mir::Body as rustc_smir[2e64afc0fcf6e61f]::rustc_smir::Stable>::stable
  16:     0x728a2fe20d57 - <rustc_smir[2e64afc0fcf6e61f]::rustc_smir::context::TablesWrapper as stable_mir[6ff77d18324bc8e5]::compiler_interface::Context>::mir_body
  17:     0x728a3012846c - <stable_mir[6ff77d18324bc8e5]::CrateItem>::body
  18:     0x728a2f2e8ee9 - rustc_driver_impl[da799d46c754177c]::pretty::print
  19:     0x728a31832a38 - rustc_interface[56d312c05374fed]::interface::run_compiler::<(), rustc_driver_impl[da799d46c754177c]::run_compiler::{closure#0}>::{closure#1}
  20:     0x728a3171c387 - std[4908e3f2635a52b0]::sys::backtrace::__rust_begin_short_backtrace::<rustc_interface[56d312c05374fed]::util::run_in_thread_with_globals<rustc_interface[56d312c05374fed]::util::run_in_thread_pool_with_globals<rustc_interface[56d312c05374fed]::interface::run_compiler<(), rustc_driver_impl[da799d46c754177c]::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}, ()>
  21:     0x728a3171c822 - <<std[4908e3f2635a52b0]::thread::Builder>::spawn_unchecked_<rustc_interface[56d312c05374fed]::util::run_in_thread_with_globals<rustc_interface[56d312c05374fed]::util::run_in_thread_pool_with_globals<rustc_interface[56d312c05374fed]::interface::run_compiler<(), rustc_driver_impl[da799d46c754177c]::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}, ()>::{closure#1} as core[94282094966b97bc]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  22:     0x728a3171ddef - std::sys::pal::unix::thread::Thread::new::thread_start::h138d3ae6a4291935
  23:     0x728a2baa339d - <unknown>
  24:     0x728a2bb2849c - <unknown>
  25:                0x0 - <unknown>

error: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: please make sure that you have updated to the latest nightly

note: rustc 1.85.0-nightly (f4f0fafd0 2024-12-13) running on x86_64-unknown-linux-gnu

note: compiler flags: -Z unpretty=stable-mir -Z dump-mir-dir=dir

query stack during panic:
end of query stack
warning: 1 warning emitted


```

</p>
</details>

<!--
query stack:

-->

```shell-session
$ rustc +nightly mre.rs --edition=2018  -Zunpretty=stable-mir
error[E0658]: async closures are unstable
 --> mre.rs:5:19
  |
5 |     let closure = async || {
  |                   ^^^^^
  |
  = note: see issue #62290 <https://github.com/rust-lang/rust/issues/62290> for more information
  = help: add `#![feature(async_closure)]` to the crate attributes to enable
  = note: this compiler was built on 2024-12-12; consider upgrading it if it is out of date
  = help: to use an async block, remove the `||`: `async {`
```

I don't think they are @matthiaskrgr

--- a/compiler/rustc_smir/src/rustc_internal/internal.rs
+++ b/compiler/rustc_smir/src/rustc_internal/internal.rs
@@ -141,6 +141,10 @@ impl RustcInternal for RigidTy {
             RigidTy::Coroutine(def, args, _mov) => {
                 rustc_ty::TyKind::Coroutine(def.0.internal(tables, tcx), args.internal(tables, tcx))
             }
+            RigidTy::CoroutineClosure(def, args) => rustc_ty::TyKind::CoroutineClosure(
+                def.0.internal(tables, tcx),
+                args.internal(tables, tcx),
+            ),
             RigidTy::CoroutineWitness(def, args) => rustc_ty::TyKind::CoroutineWitness(
                 def.0.internal(tables, tcx),
                 args.internal(tables, tcx),

--- a/compiler/rustc_smir/src/rustc_internal/mod.rs
+++ b/compiler/rustc_smir/src/rustc_internal/mod.rs
@@ -107,6 +107,10 @@ impl<'tcx> Tables<'tcx> {
         stable_mir::ty::CoroutineDef(self.create_def_id(did))
     }
 
+    pub fn coroutine_closure_def(&mut self, did: DefId) -> stable_mir::ty::CoroutineClosureDef {
+        stable_mir::ty::CoroutineClosureDef(self.create_def_id(did))
+    }
+
     pub fn alias_def(&mut self, did: DefId) -> stable_mir::ty::AliasDef {
         stable_mir::ty::AliasDef(self.create_def_id(did))
     }

--- a/compiler/rustc_smir/src/rustc_smir/convert/mir.rs
+++ b/compiler/rustc_smir/src/rustc_smir/convert/mir.rs
@@ -565,8 +565,11 @@ impl<'tcx> Stable<'tcx> for mir::AggregateKind<'tcx> {
                     tables.tcx.coroutine_movability(*def_id).stable(tables),
                 )
             }
-            mir::AggregateKind::CoroutineClosure(..) => {
-                todo!("FIXME(async_closures): Lower these to SMIR")
+            mir::AggregateKind::CoroutineClosure(def_id, generic_args) => {
+                stable_mir::mir::AggregateKind::CoroutineClosure(
+                    tables.coroutine_closure_def(*def_id),
+                    generic_args.stable(tables),
+                )
             }
             mir::AggregateKind::RawPtr(ty, mutability) => {
                 stable_mir::mir::AggregateKind::RawPtr(ty.stable(tables), mutability.stable(tables))

--- a/compiler/stable_mir/src/mir/body.rs
+++ b/compiler/stable_mir/src/mir/body.rs
@@ -5,8 +5,8 @@ use serde::Serialize;
 use crate::compiler_interface::with;
 use crate::mir::pretty::function_body;
 use crate::ty::{
-    AdtDef, ClosureDef, CoroutineDef, GenericArgs, MirConst, Movability, Region, RigidTy, Ty,
-    TyConst, TyKind, VariantIdx,
+    AdtDef, ClosureDef, CoroutineClosureDef, CoroutineDef, GenericArgs, MirConst, Movability,
+    Region, RigidTy, Ty, TyConst, TyKind, VariantIdx,
 };
 use crate::{Error, Opaque, Span, Symbol};
 
@@ -617,6 +617,9 @@ impl Rvalue {
                 AggregateKind::Coroutine(def, ref args, mov) => {
                     Ok(Ty::new_coroutine(def, args.clone(), mov))
                 }
+                AggregateKind::CoroutineClosure(def, ref args) => {
+                    Ok(Ty::new_coroutine_closure(def, args.clone()))
+                }
                 AggregateKind::RawPtr(ty, mutability) => Ok(Ty::new_ptr(ty, mutability)),
             },
             Rvalue::ShallowInitBox(_, ty) => Ok(Ty::new_box(*ty)),
@@ -633,6 +636,7 @@ pub enum AggregateKind {
     Closure(ClosureDef, GenericArgs),
     // FIXME(stable_mir): Movability here is redundant
     Coroutine(CoroutineDef, GenericArgs, Movability),
+    CoroutineClosure(CoroutineClosureDef, GenericArgs),
     RawPtr(Ty, Mutability),
 }
 

--- a/compiler/stable_mir/src/mir/pretty.rs
+++ b/compiler/stable_mir/src/mir/pretty.rs
@@ -410,6 +410,10 @@ fn pretty_aggregate<W: Write>(
             write!(writer, "{{coroutine@{}}}(", def.span().diagnostic())?;
             ")"
         }
+        AggregateKind::CoroutineClosure(def, _) => {
+            write!(writer, "{{coroutine-closure@{}}}(", def.span().diagnostic())?;
+            ")"
+        }
         AggregateKind::RawPtr(ty, mutability) => {
             write!(
                 writer,

--- a/compiler/stable_mir/src/ty.rs
+++ b/compiler/stable_mir/src/ty.rs
@@ -63,6 +63,11 @@ impl Ty {
         Ty::from_rigid_kind(RigidTy::Coroutine(def, args, mov))
     }
 
+    /// Create a new closure type.
+    pub fn new_coroutine_closure(def: CoroutineClosureDef, args: GenericArgs) -> Ty {
+        Ty::from_rigid_kind(RigidTy::CoroutineClosure(def, args))
+    }
+
     /// Create a new box type that represents `Box<T>`, for the given inner type `T`.
     pub fn new_box(inner_ty: Ty) -> Ty {
         with(|cx| cx.new_box_ty(inner_ty))
@@ -550,6 +555,7 @@ pub enum RigidTy {
     Closure(ClosureDef, GenericArgs),
     // FIXME(stable_mir): Movability here is redundant
     Coroutine(CoroutineDef, GenericArgs, Movability),
+    CoroutineClosure(CoroutineClosureDef, GenericArgs),
     Dynamic(Vec<Binder<ExistentialPredicate>>, Region, DynKind),
     Never,
     Tuple(Vec<Ty>),
@@ -738,6 +744,11 @@ crate_def! {
 crate_def! {
     #[derive(Serialize)]
     pub CoroutineDef;
+}
+
+crate_def! {
+    #[derive(Serialize)]
+    pub CoroutineClosureDef;
 }
 
 crate_def! {

--- a/compiler/stable_mir/src/visitor.rs
+++ b/compiler/stable_mir/src/visitor.rs
@@ -168,6 +168,7 @@ impl Visitable for RigidTy {
             | RigidTy::Closure(_, args)
             | RigidTy::Coroutine(_, args, _)
             | RigidTy::CoroutineWitness(_, args)
+            | RigidTy::CoroutineClosure(_, args)
             | RigidTy::FnDef(_, args) => args.visit(visitor),
             RigidTy::FnPtr(sig) => sig.visit(visitor),
             RigidTy::Dynamic(pred, r, _) => {

--- a/tests/ui/stable-mir-print/async-closure.rs
+++ b/tests/ui/stable-mir-print/async-closure.rs
@@ -1,0 +1,12 @@
+//@ compile-flags: -Z unpretty=stable-mir --crate-type lib -C panic=abort
+//@ check-pass
+//@ only-x86_64
+//@ edition: 2024
+//@ needs-unwind unwind edges are different with panic=abort
+
+pub fn foo() {
+    let y = 0;
+    let x = async || {
+        let y = y;
+    };
+}

--- a/tests/ui/stable-mir-print/async-closure.stdout
+++ b/tests/ui/stable-mir-print/async-closure.stdout
@@ -1,0 +1,90 @@
+// WARNING: This is highly experimental output it's intended for stable-mir developers only.
+// If you find a bug or want to improve the output open a issue at https://github.com/rust-lang/project-stable-mir.
+fn foo() -> () {
+    let mut _0: ();
+    let  _1: i32;
+    let  _2: {async closure@$DIR/async-closure.rs:9:13: 9:21};
+    let mut _3: &i32;
+    debug y => _1;
+    debug x => _2;
+    bb0: {
+        _1 = 0_i32;
+        _3 = &_1;
+        _2 = {coroutine-closure@$DIR/async-closure.rs:9:13: 9:21}(move _3);
+        return;
+    }
+}
+fn foo::{closure#0}(_1: &{async closure@$DIR/async-closure.rs:9:13: 9:21}) -> {async closure body@$DIR/async-closure.rs:9:22: 11:6} {
+    let mut _0: {async closure body@$DIR/async-closure.rs:9:22: 11:6};
+    let mut _2: &i32;
+    debug y => (*((*_1).0: &i32));
+    bb0: {
+        _2 = CopyForDeref(((*_1).0: &i32));
+        _0 = {coroutine@$DIR/async-closure.rs:9:22: 11:6}(_2);
+        return;
+    }
+}
+fn foo::{closure#0}::{closure#0}(_1: Pin<&mut {async closure body@$DIR/async-closure.rs:9:22: 11:6}>, _2: &mut Context<'_>) -> Poll<()> {
+    let mut _0: Poll<()>;
+    let  _3: i32;
+    let mut _4: &i32;
+    let mut _5: u32;
+    let mut _6: &mut {async closure body@$DIR/async-closure.rs:9:22: 11:6};
+    let mut _7: &mut {async closure body@$DIR/async-closure.rs:9:22: 11:6};
+    let mut _8: &mut {async closure body@$DIR/async-closure.rs:9:22: 11:6};
+    debug _task_context => _2;
+    debug y => (*((*(_1.0: &mut {async closure body@$DIR/async-closure.rs:9:22: 11:6})).0: &i32));
+    debug y => _3;
+    bb0: {
+        _6 = CopyForDeref((_1.0: &mut {async closure body@$DIR/async-closure.rs:9:22: 11:6}));
+        _5 = discriminant((*_6));
+        switchInt(move _5) -> [0: bb1, 1: bb2, otherwise: bb3];
+    }
+    bb1: {
+        _7 = CopyForDeref((_1.0: &mut {async closure body@$DIR/async-closure.rs:9:22: 11:6}));
+        _4 = CopyForDeref(((*_7).0: &i32));
+        _3 = (*_4);
+        _0 = std::task::Poll::Ready(());
+        _8 = CopyForDeref((_1.0: &mut {async closure body@$DIR/async-closure.rs:9:22: 11:6}));
+        discriminant((*_8) = 1;
+        return;
+    }
+    bb2: {
+        assert(false, `async fn` resumed after completion) -> [success: bb2, unwind unreachable];
+    }
+    bb3: {
+        unreachable;
+    }
+}
+fn foo::{closure#0}::{closure#1}(_1: Pin<&mut {async closure body@$DIR/async-closure.rs:9:22: 11:6}>, _2: &mut Context<'_>) -> Poll<()> {
+    let mut _0: Poll<()>;
+    let  _3: i32;
+    let mut _4: &i32;
+    let mut _5: u32;
+    let mut _6: &mut {async closure body@$DIR/async-closure.rs:9:22: 11:6};
+    let mut _7: &mut {async closure body@$DIR/async-closure.rs:9:22: 11:6};
+    let mut _8: &mut {async closure body@$DIR/async-closure.rs:9:22: 11:6};
+    debug _task_context => _2;
+    debug y => (*((*(_1.0: &mut {async closure body@$DIR/async-closure.rs:9:22: 11:6})).0: &i32));
+    debug y => _3;
+    bb0: {
+        _6 = CopyForDeref((_1.0: &mut {async closure body@$DIR/async-closure.rs:9:22: 11:6}));
+        _5 = discriminant((*_6));
+        switchInt(move _5) -> [0: bb1, 1: bb2, otherwise: bb3];
+    }
+    bb1: {
+        _7 = CopyForDeref((_1.0: &mut {async closure body@$DIR/async-closure.rs:9:22: 11:6}));
+        _4 = CopyForDeref(((*_7).0: &i32));
+        _3 = (*_4);
+        _0 = std::task::Poll::Ready(());
+        _8 = CopyForDeref((_1.0: &mut {async closure body@$DIR/async-closure.rs:9:22: 11:6}));
+        discriminant((*_8) = 1;
+        return;
+    }
+    bb2: {
+        assert(false, `async fn` resumed after completion) -> [success: bb2, unwind unreachable];
+    }
+    bb3: {
+        unreachable;
+    }
+}

